package/MDAnalysis/lib/lazy.py

```diff
@@ -0,0 +1,312 @@
+        # IPython tries to be too clever and constantly inspects, asking for
+        #  modules' attrs, which causes premature module loading and unesthetic
+        #  internal errors if the lazily-loaded module doesn't exist. Returning
+        #  Nones seems to satisfy those needs:
+        caller_base = _caller_name().partition('.')[0]
+        if run_from_ipython and caller_base in ('inspect', 'IPython'):
+            return None
+        _load_module(self)
+        return ModuleType.__getattribute__(self, attr)
+
+    def __setattr__(self, attr, value):
+        _load_module(self)
+        return ModuleType.__setattr__(self, attr, value)
+
+
+def _caller_name(depth=2):
+    """Returns the name of the calling namespace
+
+    """
+    # the presence of sys._getframe might be implementation-dependent.
+    # It isn't that serious if we can't get the caller's name.
+    try:
+        return sys._getframe(depth).f_globals['__name__']
+    except AttributeError:
+        return 'MDAnalysis'
+
+
+def run_from_ipython():
+    # Taken from https://stackoverflow.com/questions/5376837
+    try:
+        __IPYTHON__
+        return True
+    except NameError:
+        return False
+
+
+def import_module(modname, level='leaf'):
+    """Function allowing lazy importing of a module into the namespace
+
+    Parameters
+    ----------
+    modname : str
+         The module to import.
+    level : str, optional
+         Which submodule reference to return. Either a reference to the 'leaf'
+         module (the default) or to the 'base' module. For 'base'::
+
+            MDAnalysis = import_module("MDAnalysis.analysis.distances",
+                                       level='base')
+            # 'MDAnalysis' becomes defined in the current namespace, with
+            #  (sub)attributes 'MDAnalysis.analysis' and
+            #  'MDAnalysis.analysis.distances'.
+            # Equivalent to:
+            import MDAnalysis.analysis.distances
+
+        For 'leaf'::
+
+            distances = import_module("MDAnalysis.analysis.distances",
+                                      level='leaf')
+            # Only 'distances' becomes set in the current namespace.
+            # Equivalent to:
+            from MDAnalysis.analysis import distances
+
+    Returns
+    -------
+    module
+        The module specified by *modname*, or its base, depending on *level*.
+        The module isn't immediately imported. Instead, a
+        :class:`MDAnalysis.lib.lazy.LazyModule` instance is returned. Upon
+        access to any of its attributes, the module is finally loaded.
+
+    .. versionadded:: 0.16.2
+
+    """
+    mod = _import_module(modname, _caller_name())
+    if level == 'base':
+        return sys.modules[modname.split('.')[0]]
+    elif level == 'leaf':
+        return mod
+    else:
+        raise ValueError("Parameter 'level' must be one of ('base', 'leaf')")
+
+
+def _import_module(modname, caller_name):
+    acquire_lock()
```

Comment:
The context manager already exists, but inside `_imp`. I'd rather keep it high-level and not use `_imp`, but I already have to anyway for Python 3 support...
So I guess the answer is yes :)